pkg/monitoring/monitoring.go

```diff
@@ -0,0 +1,285 @@
+func (c *Client) createMetrics(ctx context.Context) error {
+	if err := c.createToolchainConfigsMetrics(ctx); err != nil {
+		return fmt.Errorf("unable to initialize the toolchain config runs metric: %w", err)
+	}
+	return nil
+}
+
+// createToolchainConfigsMetrics creates smetrics descriptors for the toolchain configs generation.
+func (c *Client) createToolchainConfigsMetrics(ctx context.Context) error {
+	metrics := []struct {
+		name        string
+		metricType  string
+		description string
+	}{
+		{
+			name:        "RBE Toolchain Configs Generation",
+			metricType:  mtypeToolchainConfigsGenRuns,
+			description: "Count number of times RBE Bazel C++/Java toolchain config generation completed",
+		},
+		{
+			name:        "RBE Toolchain Configs Upload",
+			metricType:  mtypeToolchainConfigsUploadRuns,
+			description: "Count number of times RBE Bazel C++/Java toolchain config upload completed",
+		},
+		{
+			name:        "RBE Toolchain Configs E2E Test",
+			metricType:  mtypeToolchainConfigsTestRuns,
+			description: "Count number of times RBE Bazel C++/Java toolchain config e2e test completed",
+		},
+	}
+
+	for _, m := range metrics {
+		md := &metric.MetricDescriptor{
+			Name: m.name,
+			Type: m.metricType,
+			Labels: []*label.LabelDescriptor{
+				{
+					Key:         "docker_image",
+					ValueType:   label.LabelDescriptor_STRING,
+					Description: "Name of the OS of the toolchain container image",
+				},
+				{
+					Key:         "result",
+					ValueType:   label.LabelDescriptor_BOOL,
+					Description: "Indicates of configs generation, upload & testing was successful",
+				},
+			},
+			MetricKind:  metric.MetricDescriptor_CUMULATIVE,
+			ValueType:   metric.MetricDescriptor_INT64,
+			Unit:        "1",
+			Description: m.description,
+			DisplayName: m.name,
+		}
+		req := &monitoringpb.CreateMetricDescriptorRequest{
+			Name:             "projects/" + c.projectID,
+			MetricDescriptor: md,
+		}
+		if _, err := c.mc.CreateMetricDescriptor(ctx, req); err != nil {
+			return fmt.Errorf("unable to create Google Cloud Monitoring Metric for %s: %v", m.name, err)
+		}
+	}
+	return nil
+}
+
+// reportCummulativeCount reports the given metric type which is expected to be of kind
+// cummulative (https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors#metrickind)
+// adding "1" to the cummulative count. Other arguments:
+// imageName: The toolchain container image for which the metric is being reported.
+// result: Indicates if the workflow was successful.
+func (c *Client) reportCummulativeCount(ctx context.Context, metricType, imageName string, result bool) error {
+	reset := &timestamp.Timestamp{
+		Seconds: c.resetTs.Unix(),
+	}
+	// For cummulative metrics, end time should be > than reset time. Thus, sleep for 2 seconds if
+	// we find less than 1s has passed since resetTs. Sleep 2s instead of 1s to avoid flakes
+	// from floating point addition errors.
+	if time.Now().Sub(c.resetTs).Seconds() < 1 {
```

Comment:
Hmm, this Won't Happen, but if it does (I'm looking at you, Daylight Saving Time!), I'd rather there not be a lurking horror: we should either change c.resetTs before the sleep (infeasible?) or print something and check in a loop.

Reply:
This is a very good point however I think I'm going to leave this as is & add a comment explaining this case isn't handled.

Rationale: If resetTs < current time, the reporting the metric will fail but the rest of the binary should be unaffected. Also, we only expect to run this tool hourly during business hours so the DST cutoff shouldn't affect our automation. Also, even if it does, at worst an alert will fire *once a year* (only when the time jumps back) and self recover within 3-4 hours (depends on the sliding window we use to calculate alert thresholds) so I don't think it's worth engineering a solution for this very rare scenario and keep the code simple instead.